### PR TITLE
CAN: CANopen: Add program download support

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
This series of patches adds program download support for CANopen by implementing the CAN in Automation (CiA) 302-3 Draft Standard Proposal (DSP).

### CODE REVIEW

File: boards/arm/frdm_k64f/board.cmake [hunk 1]
```diff
@@ -16,3 +16,4 @@ board_runner_args(pyocd "--target=k64f")
 include(${ZEPHYR_BASE}/boards/common/pyocd.board.cmake)
 include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
 include(${ZEPHYR_BASE}/boards/common/openocd.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/canopen.board.cmake)
```
Comment (henrikbrixandersen) [inline boards/arm/frdm_k64f/board.cmake]: @mbolivar I would rather have included this in the sample `CMakeLists.txt` file, but that results in an error from `west flash --runner canopen`.

**Comment by Laczen** [comment]
@henrikbrixandersen, nice work. Maybe the can_dfu could be made somewhat more versatile by introducing a dfu api (set of callback routines) that allows you to define different flash routines. An example for legacy nordic ble dfu:
```
** @brief NRF DFU callbacks structure. */
struct nrf_dfu_cb {
	/** Attribute start callback (optional)
	 *
	 *  @param img_type   The image type to update
	 *
	 *  @return 0 on success, -errno code on error
	 */
	int (*start)(enum dfu_image_type img_type);

	/** Attribute receive callback (optional)
	 *
	 *  @param offset   image offset to update/write
	 *  @param data     pointer to img data chunk
	 *  @param len      img data chunk size
	 *
	 *  @return 0 on success, -errno code on error
	 */
	int (*receive)(u32_t offset, const u8_t *data, size_t len);

	/** Attribute validate callback (optional)
	 *
	 *  @param crc   expected updated image CRC16
	 *
	 *  @return 0 on success, -errno code on error
	 */
	int (*validate)(u16_t crc);

	/** Attribute reset_to_app callback (optional)
	 *
	 * Reset system to updated image.
	 */
	void (*reset_to_app)(void);

	/** Attribute reset_to_boot callback (optional)
	 *
	 * Reset system to bootloader mode for update.
	 * Can be NULL if already in bootloader mode.
	 */
	void (*reset_to_boot)(void);
};
```
These routines are then registered to the can_dfu so that it knows what routine to call.

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen, nice work. Maybe the can_dfu could be made somewhat more versatile by introducing a dfu api (set of callback routines) that allows you to define different flash routines. An example for legacy nordic ble dfu:

@Laczen Thank you for reviewing.

> These routines are then registered to the can_dfu so that it knows what routine to call.

I'm taking you're suggesting to one common, Zephyr API for registering DFU callbacks? And then have an MCUboot specific implementation, but with the ability to add support for either other bootloaders or applications-specific routines? Where would the default implementations of these functions live? I am not sure I can lift the task of adding a complete DFU callback framework to Zephyr as part of this PR...

**Comment by Laczen** [comment]
> > @henrikbrixandersen, nice work. Maybe the can_dfu could be made somewhat more versatile by introducing a dfu api (set of callback routines) that allows you to define different flash routines. An example for legacy nordic ble dfu:
> 
> @Laczen Thank you for reviewing.
> 
> > These routines are then registered to the can_dfu so that it knows what routine to call.
> 
> I'm taking you're suggesting to one common, Zephyr API for registering DFU callbacks? And then have an MCUboot specific implementation, but with the ability to add support for either other bootloaders or applications-specific routines? Where would the default implementations of these functions live? I am not sure I can lift the task of adding a complete DFU callback framework to Zephyr as part of this PR...

I agree that addind a complete dfu callback framework should not be part of this PR. I would only define some callbacks that are needed by the can_dfu (receive, read, status, flush, confirm,...). It is the can_dfu that defines the requirements. These routines are then provided when initializing the can dfu. There default implementation is NULL, if you call them they would do nothing.

It would be very advantageous if there is a need to upgrade multiple images (which I think there is already), the can_dfu would only change a little (if any). Also it allows easy security enforcement by e.g. not providing a read routine and thereby disabling readout of firmware.

BTW, the same is valid for the usb dfu.

File: boards/arm/frdm_k64f/board.cmake (hunk 1 above)
Comment (mbolivar) [reply boards/arm/frdm_k64f/board.cmake]: >  I would rather have included this in the sample `CMakeLists.txt` file

I can see why.

I guess the feature request is to allow applications to define their own runners and configure arguments for them, right?

Comment (henrikbrixandersen) [reply boards/arm/frdm_k64f/board.cmake]: Precisely.

Comment (mbolivar) [reply boards/arm/frdm_k64f/board.cmake]: https://github.com/zephyrproject-rtos/zephyr/issues/23452 to track this.

I want to finish a few other things before looping back to new runner code, though.

**Comment by henrikbrixandersen** [comment]
The `Zephyr West Comands Tests` workflows are failing due to the added python dependencies not met by the CI docker image.

**Comment by henrikbrixandersen** [comment]
I would appreciate some feedback on this. We cannot handle it out-of-tree since west does not support out-of-tree runners yet.

**Comment by henrikbrixandersen** [comment]
> Is this lack a blocker for you? I haven't forgotten this, it's just lower down on the priority list at the moment. I could prioritize if it's a blocker.

@mbolivar-nordic Well, the best solution for us would be to have this PR merged, but it has been sitting here without any reviews for one and a half month, so... The second best solution would be if we could maintain the CANopen DFU and west runner out-of-tree.

**Review by mbolivar-nordic — COMMENTED** [review]
Can you add a couple of basic test cases for the canopen_program runner? I think I am fine with the west parts if so. I'm not asking for exhaustive testing like we have for nrfjprog, just some basic sanity checks for no-argument `west flash` style invocations I can lean on when refactoring

**Comment by henrikbrixandersen** [comment]
> Can you add a couple of basic test cases for the canopen_program runner? I think I am fine with the west parts if so. I'm not asking for exhaustive testing like we have for nrfjprog, just some basic sanity checks for no-argument `west flash` style invocations I can lean on when refactoring

@mbolivar-nordic Thank. I have added a number of test cases for the CANopen program download runner.

**Review by alexanderwachter — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
@nashif CI is failing to to added Python requirements in this PR. Any way to work-around that to get CI running?

**Review by mbolivar-nordic — APPROVED** [review]
West parts LGTM

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**